contracts/contracts/OpenAiChatGptVision.sol

```diff
@@ -0,0 +1,181 @@
+            model : "gpt-4-turbo",
+            frequencyPenalty : 21, // > 20 for null
+            logitBias : "", // empty str for null
+            maxTokens : 1000, // 0 for null
+            presencePenalty : 21, // > 20 for null
+            responseFormat : "{\"type\":\"text\"}",
+            seed : 0, // null
+            stop : "", // null
+            temperature : 10, // Example temperature (scaled up, 10 means 1.0), > 20 means null
+            topP : 101, // Percentage 0-100, > 100 means null
+            tools : "",
+            toolChoice : "", // "none" or "auto"
+            user : "" // null
+        });
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "Caller is not owner");
+        _;
+    }
+
+    modifier onlyOracle() {
+        require(msg.sender == oracleAddress, "Caller is not oracle");
+        _;
+    }
+
+    function setOracleAddress(address newOracleAddress) public onlyOwner {
+        oracleAddress = newOracleAddress;
+        emit OracleAddressUpdated(newOracleAddress);
+    }
+
+    function startChat(string memory message, string memory imageUrl) public returns (uint i) {
+        ChatRun storage run = chatRuns[chatRunsCount];
+
+        run.owner = msg.sender;
+        IOracle.Message memory newMessage = IOracle.Message({
+            role: "user",
+            content: new IOracle.Content[](2)
+        });
+        newMessage.content[0] = IOracle.Content({
+            contentType: "text",
+            value: message
+        });
+        newMessage.content[1] = IOracle.Content({
+            contentType: "image_url",
+            value: imageUrl
+        });
+        run.messages.push(newMessage);
+        run.messagesCount = 1;
+
+        uint currentId = chatRunsCount;
+        chatRunsCount = chatRunsCount + 1;
+
+        IOracle(oracleAddress).createOpenAiLlmCall(currentId, config);
+        emit ChatCreated(msg.sender, currentId);
+
+        return currentId;
+    }
+
+    function onOracleOpenAiLlmResponse(
+        uint runId,
+        IOracle.OpenAiResponse memory response,
+        string memory errorMessage
+    ) public onlyOracle {
+        ChatRun storage run = chatRuns[runId];
+        require(
+            keccak256(abi.encodePacked(run.messages[run.messagesCount - 1].role)) == keccak256(abi.encodePacked("user")),
+            "No message to respond to"
+        );
+
+        if (!compareStrings(errorMessage, "")) {
+            IOracle.Message memory newMessage = IOracle.Message({
+                role: "assistant",
+                content: new IOracle.Content[](1)
+            });
+            newMessage.content[0].contentType = "text";
+            newMessage.content[0].value = errorMessage;
+            run.messages.push(newMessage);
+            run.messagesCount++;
+        } else {
+            if (compareStrings(response.content, "")) {
+                IOracle(oracleAddress).createFunctionCall(runId, response.functionName, response.functionArguments);
```

Comment:
Can have function call here? And in `onOracleFunctionResponse`?

Reply:
I will remove a function call, since no "tools" are configured.